oasislmf/preparation/gul_inputs.py

```diff
@@ -273,83 +283,94 @@ def get_gul_input_items(
+            for generic_col, cov_col in zip(cov_type_term_cols, cov_type_terms) if generic_col in gul_inputs_df.columns
         }
         column_mapping_dict[tiv_col] = 'tiv'
-        cov_type_group.rename(columns=column_mapping_dict, inplace=True, copy=False)
-        cov_type_group.loc[:, ['tiv'] + list(column_mapping_dict.values())]
-        terms_found.update(column_mapping_dict.values())
+
+        cols_by_cov_type[cov_type] = {
+            'to_drop': other_cov_types + other_cov_type_term_cols,
+            'is_bi_coverage': is_bi_coverage,
+            'column_mapping_dict': column_mapping_dict
+        }
+
+    # coverage unpacking and disaggregation loop:
+    #  - one row representing N coverages is being transformed to N rows, one per coverage.
+    #  - if NumberOfBuildings > 1, on top of unpacking the coverage, it performs the disaggregation of the items
+    #    by repeating the rows `NumberOfBuildings` times and assigning to each row a unique `disagg_id`` number,
+    #    useful for generating `item_id` later.
+    #  - group the rows in the GUL inputs table by coverage type
+    #  - set the IL terms (and BI coverage boolean) in each group and update the corresponding frame section in the GUL inputs table
+    gul_inputs_reformatted_chunks = []
+    terms_found = set()
+    last_disagg_id = 1
+    for (number_of_buildings, cov_type), cov_type_group in gul_inputs_df.groupby(by=['NumberOfBuildings', 'coverage_type_id'], sort=True):
+        # drop columns corresponding to other cov types
+        cov_type_group.drop(
+            columns=cols_by_cov_type[cov_type]['to_drop'],
+            errors="ignore",  # Ignore if any of these cols don't exist
+            inplace=True
+        )
+
+        # check if coverage type is "bi"
+        cov_type_group['is_bi_coverage'] = cols_by_cov_type[cov_type]['is_bi_coverage']
+
+        cov_type_group.rename(columns=cols_by_cov_type[cov_type]['column_mapping_dict'], inplace=True, copy=False)
         cov_type_group['coverage_type_id'] = cov_type
-        gul_inputs_reformatted_chunks.append(cov_type_group)
-    # Concatenate chunks. Sort by index to preserve item_id order in generated outputs compared
-    # to original code.
-    gul_inputs_df = pd.concat(gul_inputs_reformatted_chunks)
-    gul_inputs_df[list(terms_found)].fillna(0., inplace=True)
-    gul_inputs_df = gul_inputs_df.sort_index().reset_index(drop=True)
-    term_int_found = list(set(gul_inputs_df.columns).intersection(set(term_cols_ints + terms_ints)))
-    gul_inputs_df[term_int_found] = gul_inputs_df[term_int_found].fillna(0)
-    # Set default values and data types for BI coverage boolean, TIV, deductibles and limit
+        terms_found.update(cols_by_cov_type[cov_type]['column_mapping_dict'].values())
+
+        # if NumberOfBuildings == 0: still add one entry
+        disagg_df = pd.concat([cov_type_group] * max(number_of_buildings, 1), ignore_index=True)
```

Comment:
I think if you ignore index here you cannot sort index later on in:
    gul_inputs_df = (
        pd.concat(gul_inputs_reformatted_chunks)
        .fillna(value={c: 0 for c in terms_found})
        .sort_index()
        .reset_index(drop=True)
        .fillna(value={c: 0 for c in set(gul_inputs_df.columns).intersection(set(term_cols_ints + terms_ints))})
    )